packages/api/src/utils/client/method.ts

```diff
@@ -0,0 +1,50 @@
+import {mapValues} from "@lodestar/utils";
+import {Endpoint, HasOnlyOptionalProps, RouteDefinition, RouteDefinitions} from "../types.js";
+import {compileRouteUrlFormatter} from "../urlFormat.js";
+import {IHttpClient} from "./httpClient.js";
+import {ApiRequestInit} from "./request.js";
+import {ApiResponse} from "./response.js";
+
+export type ApiClientMethod<E extends Endpoint> = E["args"] extends void
+  ? (init?: ApiRequestInit) => Promise<ApiResponse<E>>
+  : HasOnlyOptionalProps<E["args"]> extends true
+    ? (args?: E["args"], init?: ApiRequestInit) => Promise<ApiResponse<E>>
+    : (args: E["args"], init?: ApiRequestInit) => Promise<ApiResponse<E>>;
+
+export type ApiClientMethods<Es extends Record<string, Endpoint>> = {[K in keyof Es]: ApiClientMethod<Es[K]>};
+
+export function createApiClientMethod<E extends Endpoint>(
+  definition: RouteDefinition<E>,
+  client: IHttpClient,
+  operationId: string
+): ApiClientMethod<E> {
+  const urlFormatter = compileRouteUrlFormatter(definition.url);
+  const definitionExtended = {
+    ...definition,
+    urlFormatter,
+    operationId,
+  };
+
+  // If the request args is void, then completely remove the args parameter
+  if (
+    definition.req.schema.body === undefined &&
+    definition.req.schema.params === undefined &&
+    definition.req.schema.query === undefined
+  ) {
+    return (async (init?: ApiRequestInit) => {
+      return client.request(definitionExtended, undefined, init);
+    }) as ApiClientMethod<E>;
+  }
+  return async (args?: E["args"], init?: ApiRequestInit) => {
+    return client.request(definitionExtended, args ?? {}, init);
+  };
+}
+
+export function createApiClientMethods<Es extends Record<string, Endpoint>>(
+  definitions: RouteDefinitions<Es>,
+  client: IHttpClient
+): ApiClientMethods<Es> {
+  return mapValues(definitions, (definition, operationId) => {
+    return createApiClientMethod(definition, client, operationId as string);
+  }) as unknown as ApiClientMethods<Es>;
```

Comment:
I agree we should try to avoid using `unknown as`, I previously looked at this already and gave it another attempt today but I don't see what's the issue, the types look good to me, seems to be some issue how `mapValues` remaps the types, the type error is pretty strange `Type 'Es[keyof Es]' is not comparable to type 'Es[K]'`

We have the same type cast on unstable branch
https://github.com/ChainSafe/lodestar/blob/c63509800c61eb44e59d41187bb272c0d1e1e462/packages/api/src/utils/client/client.ts#L124

My guess is there is some issue with the typing of `mapValues` that makes the translation incorrect but not sure how to fix it right now, maybe you could take a look and help with that. If it's related to `mapValues` type might be better to fix on unstable branch first

Reply:
Seems you have some constraints inside `ApiClientMethods` which is not aligned here. 

As you are doing this whole refactoring in this branch so I suggest to fix this types issues as well. If not now, then create an issue for later on.

Reply:
If I don't use `mapValues` I can get rid of the `as unknown`, I don't think there is a type issue in `ApiClientMethods`, or at least I don't see it.
```ts
export function createApiClientMethods<Es extends Record<string, Endpoint>>(
  definitions: RouteDefinitions<Es>,
  client: IHttpClient
): ApiClientMethods<Es> {
  const output = {} as ApiClientMethods<Es>;
  for (const [key, value] of Object.entries(definitions)) {
    output[key as keyof Es] = createApiClientMethod(value, client, key);
  }
  return output;
}
```

I created an issue to further investigate this
- https://github.com/ChainSafe/lodestar/issues/6832